[nnc] Use BufHandle in loopnest.cache_accesses python API

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59006 [nnc] Use BufHandle in loopnest.cache_accesses python API**

Test in issue #59002

Differential Revision: [D28714829](https://our.internmc.facebook.com/intern/diff/D28714829)

huiguoo committed:
[nnc] Use BufHandle in loopnest.cache_accesses python API
  Test in issue #59002

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28714829).

navahgar commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -494,10 +494,12 @@ void initTensorExprBindings(PyObject* module) {
           "cache_accesses",
           [](LoopNest& self,
-             const Buf* producer,
+             const BufHandle& producer,
```
Why is this fixing the error? Why can't we use `Buf*` here?

huiguoo replied:
Oh, this PR does not fix the error. It is needed to reproduce the error. Specifically, `Buf` has not been added in python bindings but `BufHandle` has. So we have to modify its API to use `cache_accesses` transformation in python.

navahgar replied:
Aah okay. Sounds good.

ZolotukhinM replied:
I wonder why we pass buf handles as `const BufHandle*` in other places in this file. @navahgar @Chillee, any ideas about that? Going with the reference instead of the pointer seems right to me.

navahgar replied:
Yeah, that is definitely not the right way. We should either have `const BufHandle&` if it will not be modified or a `BufHandle&` (or even pass by value) if it will be modified.

huiguoo replied:
I'll start a PR for that.

huiguoo committed:
Update on "[nnc] Use BufHandle in loopnest.cache_accesses python API"


Test in issue #59002

Differential Revision: [D28714829](https://our.internmc.facebook.com/intern/diff/D28714829)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28714829).
